### ISSUE TITLE
CamGear: self.frame always gives the same frame when camera is busy

### ISSUE THREAD

**Issue by agjunyent** [abhiTronix/vidgear#220]
I'd like to suggest to set `self.frame = None` before [(grabbed, frame) = self.stream.read()](https://github.com/abhiTronix/vidgear/blob/eb19b9ce96061c09b5fe4c3ee63d84c940a19af2/vidgear/gears/camgear.py#L359) in order to improve the ability to synchronize two cameras with different FPS and functionalities.

### Context

I'm using vidgear on a Raspi4 with 2 USB cameras: RGB camera and Thermal Camera. The idea is to track temperature in a zone, and match it to different objects using ML (out of scope). I was using pure OpenCV till now, but I saw your code and wanted to give it a try.

My problem is that while the RGB camera has 15 or 30fps modes, the Thermal Camera has 9fps fixed mode. Also every once in a while, the Thermal Camera stops sending frames while it autocalibrates, to avoid drift over time of temperature values. 

During this time, when I was using OpenCV, cap.read() hangs, so I was able to "wait" in the RGB camera until the thermal camera responded, and comparing timestamps I was able to synchronize. With your code, the only way to check for frames is with `self.frame is not None` (as far as I see in your docs). So when the Thermal camera stops, no new frames are sent, and the RGB camera advances while the Thermal camera gives always the same frame, making my ML go crazy as the synchronization is wrong.

If you add `self.frame = None` BEFORE  [(grabbed, frame) = self.stream.read()](https://github.com/abhiTronix/vidgear/blob/eb19b9ce96061c09b5fe4c3ee63d84c940a19af2/vidgear/gears/camgear.py#L359) , while the Thermal camera is waiting, the "main" code will see it and wait, as no new frames are grabbed.

Will this change the existing VidGear APIs? How?
No

How can it benefit other users?
Same as me: if two cameras have different FPS, or one camera is slower, etc.

### Your Current Environment

* VidGear version:  0.2.1
* Branch:  pip3
* Python version:  Python 3.7.3
* Operating System and version: RaspberryPI OS 5.10.17-v7l+ #1421 SMP Thu May 27 14:00:13 BST 2021 armv7l GNU/Linux

**Comment by abhiTronix** [comment]
@agjunyent Thanks for opening this issue. But I'm not getting the gist of the actual problem, Can you provide code or related diagram, So that we can be on same page? 

VidGear is pretty good at handling frames synchronization , you can see this test here https://github.com/abhiTronix/vidgear/blob/eb19b9ce96061c09b5fe4c3ee63d84c940a19af2/vidgear/tests/videocapture_tests/test_camgear.py#L109-L141 

> During this time, when I was using OpenCV, cap.read() hangs, so I was able to "wait" in the RGB camera until the thermal camera responded, and comparing timestamps I was able to synchronize

Actually VidGear's CamGear API does the same thing. It also blocks the thread until its queue is full i.e. when you stop extracting frames from its queue. 

> With your code, the only way to check for frames is with self.frame is not None (as far as I see in your docs).

**Nope, that's the wrong way and not mentioned anywhere in docs, if does then please let me know.** The `read()` function itself the right way because it hangs the thread unless new frame is received from source just like OpenCV, also see docs: https://abhitronix.github.io/vidgear/latest/bonus/reference/camgear/#vidgear.gears.camgear.CamGear.read. You just need to use same code as mentioned in docs: https://abhitronix.github.io/vidgear/latest/help/camgear_faqs/#how-to-synchronize-between-two-cameras and it will automatically blocks until new frame is received.

> If you add self.frame = None BEFORE (grabbed, frame) = self.stream.read() , while the Thermal camera is waiting, the "main" code will see it and wait, as no new frames are grabbed. Will this change the existing VidGear APIs? How?

Actually adding this will do nothing but cause more trouble because of the Threaded nature of VidGear's VideoCapture APIs. I've dealt with synchronization issue long ago.

**Comment by abhiTronix** [comment]
Reopened as waiting for response from the author.

**Comment by agjunyent** [comment]
Hi thanks for your response.

I thought that CamGear retrieves the frames on a [separate thread](https://github.com/abhiTronix/vidgear/blob/eb19b9ce96061c09b5fe4c3ee63d84c940a19af2/vidgear/gears/camgear.py#L346-L415), and when you use `stream.read()` you are only getting the last frame that the thread has retrieved. If the camera freezes for some reason for 2-3 seconds, you are getting the same frame during that time, without a way of knowing it (this is what happens on my Thermal Camera).

Also the queued mode is disabled for USB streams as per [This THREADED_QUEUE_MODE attribute does NOT work with Live feed, such as Camera Devices/Modules.](https://abhitronix.github.io/vidgear/v0.2.1-stable/bonus/TQM/#manually-disabling-threaded-queue-mode).

I think I'm missing something somewhere, in the docs or on your code. I'll keep trying and investigating

**Comment by abhiTronix** [comment]
> During this time, when I was using OpenCV, cap.read() hangs, so I was able to "wait" in the RGB camera until the thermal camera responded, and comparing timestamps I was able to synchronize. With your code, the only way to check for frames is with self.frame is not None (as far as I see in your docs). So when the Thermal camera stops, no new frames are sent, and the RGB camera advances while the Thermal camera gives always the same frame, making my ML go crazy as the synchronization is wrong.
>Also the queued mode is disabled for USB streams as per This THREADED_QUEUE_MODE attribute does NOT work with Live feed, such as Camera Devices/Modules..

@agjunyent Ohk Now, I understand your problem and it all makes sense. Correct if I'm wrong:

You got a camera livefeed for which [Threaded Mode is disabled](https://abhitronix.github.io/vidgear/latest/bonus/TQM/#manually-disabling-threaded-queue-mode) and it gets busy for sometimes, thereby the `self.frame` in `read()` 

https://github.com/abhiTronix/vidgear/blob/eb19b9ce96061c09b5fe4c3ee63d84c940a19af2/vidgear/gears/camgear.py#L415

keeps on returning previous frame. While source thread at `update()` is blocked at `(grabbed, frame) = self.stream.read()`.

https://github.com/abhiTronix/vidgear/blob/eb19b9ce96061c09b5fe4c3ee63d84c940a19af2/vidgear/gears/camgear.py#L359

And this leads to ML programs behave erratically.


**That's sound like a problem to me. Thank you bringing this to my attention, I'll try to fix asap. Please stay tuned, I'll notify you here once it is done.**

**Comment by agjunyent** [comment]
Hi @abhiTronix. Thanks again for your response.

Exactly! That's what is happening. The camera that is not busy keeps sending frames, losing syncronization with the one that is busy "recalibrating" (in my case). This makes the ML program behave wrong as it needs both frames to be syncronized, or it will compare a "newer" frame from the RGB camera with an old frame of the Thermal camera, and if something is moving in the picture, it will not match.

Adding some way of checking, for example, the last time that a frame was grabbed, you can compare timestamps between cameras, and if the difference is too high, discard and wait for new frames. This is how I was doing it in my pure OpenCV code, setting a `self.last_frame_time = time.time()` just after `cap.read()`, so if `cap.read()` hangs, the timestamp is not updated. Then in my "main" code I was comparing frame_times between the two cameras, and if the difference was too big, discard.

Thanks!! I'll stay tuned

**Comment by abhiTronix** [comment]
@agjunyent Thanks for you patience, the problem might be fixed with latest commit. Please test it on your setup by installing `development` branch as follows:
```sh
 # clone the repository and get inside
  git clone https://github.com/abhiTronix/vidgear.git && cd vidgear

  # checkout the latest development branch
  git checkout development

  # install normally
  pip install .

  # OR install with asyncio support
  pip install .[asyncio]
```

After installing you can now use `read()` as block function that self-blocks when your camera is busy, just like  `(grabbed, frame) = self.stream.read()`, for a given timeout value _(only if specified, otherwise returns `None`)_. Also, FYI CamGear offers [`THREAD_TIMEOUT`](https://abhitronix.github.io/vidgear/v0.2.2-dev/gears/camgear/advanced/source_params/#exclusive-camgear-parameters) exclusive flag to set this timeout value(in seconds)). 

**Kindly test it as soon as possible and share your results here**. Goodluck!

**Comment by agjunyent** [comment]
Hi @abhiTronix Thanks for your fast update!! I'll give it a try tomorrow and report back results as soon as possible.

**Comment by abhiTronix** [comment]
@agjunyent any updates?

**Comment by agjunyent** [comment]
@abhiTronix being tested right now. I want to leave it running for some time to see if the desyncronization happens. I will come back tomorrow with final result.

Thanks for your time!

**Comment by abhiTronix** [comment]
@agjunyent ???

**Comment by agjunyent** [comment]
Hi @abhiTronix sorry for the late response. I've been very busy today.

Just checked and the syncronization is still perfect. Thanks!!

Your solution solved the problem, at least for me! Amazing library. I'm moving all my camera code to vidgear. Thanks!

**Comment by abhiTronix** [comment]
@agjunyent Thanks man and kudos!!!

**Comment by abhiTronix** [comment]
**Successfully Resolved and Merged in commit https://github.com/abhiTronix/vidgear/commit/efdb1e7a17178668eb07f939fb3ea96c115ee483**
